lib/puppet/context/trusted_information.rb

```diff
@@ -46,11 +41,21 @@ def initialize(authenticated, certname, extensions, external = {})
     end
     @hostname = hostname.freeze
     @domain = domain.freeze
-    @external = external.freeze
+    @external = external.eql?(UnevaluatedExternal) ? external : external.freeze
+  end
+
+  # Additional external facts loaded through `trusted_external_command`.
+  #
+  # @return [Hash]
+  def external
+    if @external.eql?(UnevaluatedExternal)
+      @external = retrieve_trusted_external(@certname).freeze
+    end
+    @external
   end
 
   def self.remote(authenticated, node_name, certificate)
-    external = retrieve_trusted_external(node_name)
+    external = UnevaluatedExternal
```

Comment:
For lazy evaluation we usually use `proc`. So here define:

```ruby
external = proc { retrieve_trusted_external(certname) }
```

and then in the getter method:

```ruby
def external
  if @external.is_a?(Proc)
    @external = @external.call.freeze
  end
  @external
end
```

Reply:
Okay, I can do that.

Style-wise, would it be best to do that twice, once in `self.remote` and once in `self.local`, or to make it a method? e.g.

```ruby
def lazy_external(node_name)
  proc { retrieve_trusted_external(node_name) }
end
```

Also, I can assume from this that there is no concern over external users passing a Proc into `new()`, yes?

Reply:
Updated PR to use `proc`, went with just defining a proc directly in both places one is needed, `self.remote` and `self.local`.